### ISSUE TITLE
Convarianve / Contrvariance for generics?

### ISSUE THREAD

**Issue by MaxGraey** [JSMonk/hegel#273]
Seems it doesn't support in hegel?
```ts
class Base {}
class Foo extends Base {}

const fooArr: Array<Foo> = [new Foo()];
const baseArr: Array<Base> = fooArr;
```
[playground](https://hegel.js.org/try#MYGwhgzhAEBCkFNoG8C+AoUkYDED2e0CAHgC4IB2AJjPBEmupnhRKdAGYECCATrwC5ofXmACeAHnx4AfNAC80ANoUEAd2jSAFAEoAugG5mrdgCNEIoSPES6COYq54RBoA)

**Comment by raveclassic** [comment]
It would be so great if Hegel supported explicit variance annotations like it's done in Scala: https://docs.scala-lang.org/tour/variances.html

**Comment by JSMonk** [comment]
@MaxGraey, yes. Because the problem is next (I describe it in details in [Array Subtyping](https://hegel.js.org/docs/array-types#subtyping) ):
```ts
// Your code
class Base {}
class Foo extends Base { someMethod() {} }

const fooArr: Array<Foo> = [new Foo()];
const baseArr: Array<Base> = fooArr;

// Runtime error code
baseArr[1] = new Base;
fooArr[1].someMethod();
```

**Comment by MaxGraey** [comment]
Right. And runtime exception is normal behaviour for C#, Java. [Rust](https://doc.rust-lang.org/nomicon/subtyping.html#variance) allow this only for traits. But as I understand Hegel doesn't have interfaces or traits?

**Comment by trusktr** [comment]
I no runtime exception is better. :+1: We can re-write the code to fit within any limitation.

I think we can allow certain things. For example, if we are adding items to an Array<Foo> in some place that expect an Array<Base>, then _that_ should not be allowed. But, if we're only reading the values from the array, that should be perfectly fine. So I think the type system can be improved, and can be more accurate depending on what the code is trying to do.

For example,

no type error in this code:

```js
// Your code
class Base {baseMethod() {}}
class Foo extends Base { someMethod() {} }

const fooArr: Array<Foo> = [new Foo()];
const baseArr: Array<Base> = fooArr; // ok, no error here
baseArr[0].baseMethod() // ok
```

type error in this code:


```js
// Your code
class Base {baseMethod() {}}
class Foo extends Base { someMethod() {} }

const fooArr: Array<Foo> = [new Foo()];
const baseArr: Array<Base> = fooArr; // ok
baseArr.push(new Base()) // error
```

So, if the type checker sees both the definition sites and the assignment (or arg passing) sites, then it can know about how the arrays will be used. If for some reason it can not tell, then it can fall back to the current behavior. But maybe it can always tell?

**Comment by kaleb** [comment]
Is there a way to signify that an array is immutable? If so this should not be an issue for immutable arrays.

**Comment by JSMonk** [comment]
@kaleb . Yes, we have a special type called `$Immutable` which creates immutable arrays.
And with the immutable array, this case will work exactly as @trusktr described.
We currently have few bugs with assign to immutable array, but it will be fixed soon.
```typescript
class Base { baseMethod() {} }
class Foo extends Base { someMethod() {} }

const fooArr: $Immutable<Array<Foo>> = [new Foo()];
const baseArr: $Immutable<Array<Base>> = fooArr; // ok
baseArr.push(new Base()) // ERROR
```
You can try it in [Playground](https://hegel.js.org/try#MYGwhgzhAEBCkFNoG9oCNEFkEBcAWA9gCYAUAlCgL7SUBQokMAYgQdAgB44IB2RM8CElQQCAWwTZ8xclRq16BHhBzQAZqwCCAJ20AuaABIAkmLEBXHGDQgEAHh3awATzssCAPg-QAvNADaPAgA7tDu5AC6ANyKyqoYQo4GJmaW1rYOui52gghevupaulHQAPSl0AQA1rQJCI4AdAAO5hB4JEGhueQU5dAAogBKgwDyg0A).

**Comment by leunam217** [comment]
I don't think that it would bring a lot of value as you can construct variant containers with immutability.
![imagen](https://user-images.githubusercontent.com/38218036/90270416-746dd780-de5a-11ea-942b-346e4f18a68c.png)

**Comment by thecotne** [comment]
@leunam217 immutability should be part of type system and should not leak artifacts to production code

you have introduced wrapper class but it is only used for type checking

**Comment by leunam217** [comment]
@thecotne I'm not sure to understand your point. But it was just a comment saying that with the current features we can simulate covariant wrappers types. Sorry if it was confusing. Thanks for your awesome work, I think that the hegel phylosophy is the way to go.

**Comment by trusktr** [comment]
> Yes, we have a special type called `$Immutable` which creates immutable arrays.
> And with the immutable array, this case will work exactly as @trusktr described.

That's great to know! However in my example I was imagining it with a mutable array (because the mutability is useful).

What I mean in that example is that the `push` would fail because although the type of `baseArr` is `Array<Base>`, there is enough code context that Hegel can understand that the array is still of type `Array<Foo>`, and can still prevent errors.

So the error would be more like

```
baseArr.push(new Base()) // ERROR, Can not push 'Base' to 'Array<Base>' which is derived from 'Array<Foo>'
```

or something (bikesheddable wording).

In other words, if Hegel has broader context, it can prevent errors, even if the types are more generic. I don't think I've seen that in a language before (but then again, my experience with typed languages is mainly C/C++, Java, and TypeScript). Is there any other language that keeps track of a broader context like that?

**Comment by trusktr** [comment]
The context could also be lost. For example, if someone write a library, where the entire library is this:

```ts
export class Base { baseMethod() {} }

export function doSomethingWithArray(arr: Array<Base>) {
  baseArr.push(new Base()) // ok
}
```

and the author compiles that to plain JS, then in that case there is no possible context that Hegel could infer from it. So in that case, the line

```ts
baseArr.push(new Base())
```

is totally ok, there is no type error because there is no context to check against. So these sorts of error-preventing type errors can only happen when the broader context is known (otherwise, there's nothing we can do).

If the end user of that library is also using Hegel, and imports the function, then this does cause a type error:

```ts
import {doSomethingWithArray, Base} from 'the-lib'

class Foo extends Base { someMethod() {} }

const fooArray = [new Foo()]
doSomethingWithArray(fooArray) // this context causes a type error, f.e. Can not push 'Base' to 'Array<Base>' which is derived from 'Array<Foo>'
```

See what I mean? Knowing the context of the code, Hegel can prevent certain errors (preventing having a non-Foo inserted into the lib user's `Array<Foo>`), while the array is _not immutable_. The user could, for example, decide to write some different code:

```ts
import {doSomethingWithArray, Base} from 'the-lib'

class Foo extends Base { someMethod() {} }

const fooArray: Array<Base> = [new Foo()]
doSomethingWithArray(fooArray) // No error, user can made the context more generic.
```

Now the user can pass that _mutable_ array into the function, athough they may now face other limitations like now being able to call Foo methods on the items. There's trade-offs, but either way the user chooses, the errors can be prevented while still having mutable arrays.
